bundles/config/org.eclipse.smarthome.config.discovery.usbserial.test/src/test/java/org/eclipse/smarthome/config/discovery/usbserial/internal/UsbSerialDiscoveryServiceTest.java

```diff
@@ -0,0 +1,182 @@
+        service.usbSerialDeviceDiscovered(deviceInfoA);
+        verify(discoveryListener, times(1)).thingDiscovered(service, discoveryResultA);
+
+        // when only the second discovery participant supports a newly discovered device, the device is also discovered
+        UsbSerialDeviceInformation deviceInfoB = generateDeviceInfo();
+        DiscoveryResult discoveryResultB = mock(DiscoveryResult.class);
+        when(discoveryParticipantA.createResult(deviceInfoB)).thenReturn(discoveryResultB);
+        service.usbSerialDeviceDiscovered(deviceInfoB);
+        verify(discoveryListener, times(1)).thingDiscovered(service, discoveryResultB);
+    }
+
+    @Test
+    public void testDiscoveredThingsAreRemoved() {
+        UsbSerialDiscoveryService service = getService(UsbSerialDiscoveryService.class);
+
+        // register one discovery listener
+        DiscoveryListener discoveryListener = mock(DiscoveryListener.class);
+        service.addDiscoveryListener(discoveryListener);
+
+        // register one discovery participant
+        UsbSerialDiscoveryParticipant discoveryParticipant = mock(UsbSerialDiscoveryParticipant.class);
+        registerService(discoveryParticipant);
+
+        // when the discovery participant does not support a removed device, no discovery result is removed
+        when(discoveryParticipant.createResult(any())).thenReturn(null);
+        service.usbSerialDeviceRemoved(generateDeviceInfo());
+        verify(discoveryListener, never()).thingRemoved(any(), any());
+
+        // when the first discovery participant supports a removed device, the discovery result is removed
+        UsbSerialDeviceInformation deviceInfo = generateDeviceInfo();
+        ThingUID thingUID = mock(ThingUID.class);
+        when(discoveryParticipant.getThingUID(deviceInfo)).thenReturn(thingUID);
+        service.usbSerialDeviceRemoved(deviceInfo);
+        verify(discoveryListener, times(1)).thingRemoved(service, thingUID);
+    }
+
+    private void setBackgroundDiscovery(boolean status) throws IOException, InterruptedException {
+        ConfigurationAdmin configAdmin = getService(ConfigurationAdmin.class);
+        Configuration configuration = configAdmin.getConfiguration("discovery.usbserial");
+        Hashtable<String, Object> properties = new Hashtable<>();
+        properties.put(CONFIG_PROPERTY_BACKGROUND_DISCOVERY, Boolean.valueOf(status));
+        configuration.update(properties);
+
+        // wait until the configuration is actually set in the discovery service
+        waitForAssert(() -> {
+            UsbSerialDiscoveryService service = getService(UsbSerialDiscoveryService.class);
+            assertThat(service, is(not(nullValue())));
+            assertThat(service.isBackgroundDiscoveryEnabled(), is(status));
+        }, 1000, 100);
+    }
+
+    private final AtomicInteger deviceInfoCounter = new AtomicInteger(0);
+
+    public UsbSerialDeviceInformation generateDeviceInfo() {
```

Comment:
Why are you not using your `UsbSerialDeviceInformationGenerator`?

Reply:
I do not use it, because I did not want to introduce a dependency between the test fragment for the bundle `org.eclipse.smarthome.config.discovery.usbserial` and the bundle `org.eclipse.smarthome.config.discovery.usbserial.linux.sysfs` just for this simple test data generator.

Do you think it is worthwile to introduce a dependency for this?

Reply:
Since you need the code in both bundles anyway I think a dependency is better than duplicating the code.

But I would create the dependency in the other direction: Move the generator into the `org.eclipse.smarthome.config.discovery.usbserial.test` bundle and have this as a dependency in the `org.eclipse.smarthome.config.discovery.usbserial.linux.sysfs.test` bundle.

WDYT?

Reply:
I, personally, would not introduce the dependency (as the piece of code is very short and quite simple). But I don't have such a strong opinion on that, so I will change as suggested.